src/BlendingCommon/HtmlViewCompositionHelper.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Starcounter;
+using Starcounter.Internal;
+
+namespace BlendingCommon
+{
+    public static class HtmlViewCompositionHelper
+    {
+        public static List<HtmlViewComposition> GetCurrentCompositions(string mainViewUri)
+        {
+            var mainView = GetView(mainViewUri);
+            return Db.SQL<HtmlViewComposition>("SELECT c FROM Starcounter.HtmlViewComposition c WHERE c.MainView = ? AND c.NextVersion IS NULL", mainView).ToList();
+        }
+
+        public static HtmlView GetOrCreateView(string viewUri)
+        {
+            return GetView(viewUri) ?? new HtmlView { Uri = viewUri };
+        }
+
+        public static HtmlView GetView(string viewUri)
+        {
+            return Db.SQL<HtmlView>("SELECT v FROM Starcounter.HtmlView v WHERE v.Uri = ?", viewUri).FirstOrDefault();
+        }
+
+        public static string GetSafeFilename(string filename)
+        {
+
+            return string.Join("_", filename.Split(Path.GetInvalidFileNameChars()));
+
+        }
+
+        public static string GetHtmlFileName(HtmlViewComposition composition)
+        {
+            var fileName = composition.MainView?.Uri;
+
+            if (fileName == null)
+            {
+                return null;
+            }
+
+            fileName = GetSafeFilename(fileName);
+            fileName = Path.GetFileNameWithoutExtension(fileName);
+
+            foreach (var attachment in composition.Attachments)
```

Comment:
The SQL `SELECT` result is never guaranteed to be in any specific order without `ORDER BY`, thus the `GetHtmlFileName` method is producing different results for the same composition.

```suggestion
            foreach (var attachment in composition.Attachments.OrderBy(x => x.AttachedView.Uri))
```

Reply:
Wow, I didn't know that!

In your suggestion, you are letting Linq `OrderedEnumerable` sort the result set instead of Starcounter database engine. Isn't it preferable to add the `ORDER` clause within the query?

Reply:
That's right, it's better to sort it on the database level. The proper way would be to change `HtmlViewComposition.Attachments` property:

```diff
-public IEnumerable<HtmlViewAttachment> Attachments => Db.SQL<HtmlViewAttachment>("SELECT a FROM Starcounter.HtmlViewAttachment a WHERE a.Composition = ?", this);
+public IQueryable<HtmlViewAttachment> Attachments => DbLinq.Objects<HtmlViewAttachment>().Where(x => x.Composition == this);
```

But since we don't want to depend on [`Starcounter.Linq`](https://github.com/Starcounter/Starcounter.Linq) in Level1, the easy solution is to use full query instead of the navigation property:

```cs
DbLinq.Objects<HtmlViewAttachment>().Where(x => x.Composition == Composition).OrderBy(x => x.AttachedView.Uri);
```

Reply:
P.S. In this particular case, the performance of the database and the .NET CLR sorting will be the same.

Reply:
Thanks! I have appllied your advice in the commit https://github.com/Starcounter/Blending/pull/320/commits/af320e2214c255fdd489e0de19f164c13e24e154.

To apply it in Starcounter, I would need to change `Level1` and I decided not to.